### PR TITLE
Node Explorer: poll for peer/tailnet updates

### LINKED ISSUE #122: Node Explorer: poll for updates from status
We should not rely on the user clicking the refresh button and should poll at a set interval (5 second default).

### PR DESCRIPTION
**Pull request by tendstofortytwo**
Fixes #122.

### CODE REVIEW

File: src/node-explorer-provider.ts [hunk 1]
```diff
@@ -56,12 +58,81 @@ export class NodeExplorerProvider
...
     return element;
   }
 
+  private async getPeers() {
+    this.previousStatus = this.currrentStatus;
+    this.currrentStatus = await this.ts.getPeers();
+    return this.currrentStatus;
+  }
+
+  private async diffRelay() {
```
Comment (tendstofortytwo) [inline src/node-explorer-provider.ts:74]: I'm not sure if this is the best way to implement this function -- feedback appreciated!

File: src/node-explorer-provider.ts [hunk 2]
```diff
@@ -56,12 +58,81 @@ export class NodeExplorerProvider
...
+
+      if (status.CurrentTailnet.MagicDNSEnabled !== prevStatus.CurrentTailnet.MagicDNSEnabled) {
+        return true;
+      }
+
+      if (status.CurrentTailnet.MagicDNSSuffix !== prevStatus.CurrentTailnet.MagicDNSSuffix) {
+        return true;
+      }
+
+      for (let i = 0; i < status.PeerGroups.length; ++i) {
```
Comment (tylersmalley) [inline src/node-explorer-provider.ts:97]: @marwan-at-work, do you know if this order is consistent? I remember you doing some digging related to that.

File: src/node-explorer-provider.ts [hunk 3]
```diff
@@ -56,12 +58,81 @@ export class NodeExplorerProvider
...
+        if (status.PeerGroups[i].Name !== prevStatus.PeerGroups[i].Name) {
+          return true;
+        }
+        if (
+          !status.PeerGroups[i].Peers.every((p, j) => p.ID === prevStatus.PeerGroups[i].Peers[j].ID)
```
Comment (tylersmalley) [inline src/node-explorer-provider.ts:102]: @marwan-at-work, do you know if this order is consistent? I remember you doing some digging related to that.

Comment (tylersmalley) [inline src/node-explorer-provider.ts:102]: Before iterating, we should probably compare the lengths first.

Comment (tylersmalley) [reply src/node-explorer-provider.ts:102]: Also, can we check to see if `Online` for the node changes?

File: src/node-explorer-provider.ts (hunk 1 above)
Comment (tylersmalley) [reply src/node-explorer-provider.ts:74]: I think this is a sensible approach. 👍

**Review by tylersmalley — APPROVED** [review]
One small change, then LGTM.

File: src/node-explorer-provider.ts [hunk 4]
```diff
@@ -56,12 +58,81 @@ export class NodeExplorerProvider
...
+    this.pollForUpdates();
   }
 
   getTreeItem(element: PeerBaseTreeItem): vscode.TreeItem {
     return element;
   }
 
+  private async getPeers() {
+    this.previousStatus = this.currrentStatus;
+    this.currrentStatus = await this.ts.getPeers();
```
Comment (tylersmalley) [inline src/node-explorer-provider.ts:70]: typo
```suggestion
    this.currentStatus = await this.ts.getPeers();
```

File: src/node-explorer-provider.ts [hunk 5]
```diff
@@ -56,12 +58,81 @@ export class NodeExplorerProvider
...
+    return this.currrentStatus;
+  }
+
+  private async diffRelay() {
+    const status = await this.getPeers();
+    const prevStatus = this.previousStatus;
+
+    if (prevStatus) {
```
Comment (tylersmalley) [inline src/node-explorer-provider.ts:78]: Can return early if there is no `prevStatus` and remove the nested if block

**Pull request merged**